lti_consumer/lti_consumer.py

```diff
@@ -0,0 +1,829 @@
+            actions initially weren't required.
+    B) LTI 2.0 Result Service JSON REST endpoint
+       (http://www.imsglobal.org/lti/ltiv2p0/uml/purl.imsglobal.org/vocab/lis/v2/outcomes/Result/service.html)
+        a.) Discovery of all such LTI http endpoints for a course.  External tools GET from this discovery
+            endpoint and receive URLs for interacting with individual grading units.
+            (see lms/djangoapps/courseware/views.py:get_course_lti_endpoints)
+        b.) GET, PUT and DELETE in LTI Result JSON binding
+            (http://www.imsglobal.org/lti/ltiv2p0/mediatype/application/vnd/ims/lis/v2/result+json/index.html)
+            for a provider to synchronize grades into edx-platform.  Reading, Setting, and Deleteing
+            Numeric grades between 0 and 1 and text + basic HTML feedback comments are supported, via
+            GET / PUT / DELETE HTTP methods respectively
+"""
+
+import logging
+import bleach
+import re
+import json
+import urllib
+
+from datetime import datetime
+from collections import namedtuple
+from webob import Response
+
+from xblock.core import String, Scope, List, XBlock
+from xblock.fields import Boolean, Float, Integer
+from xblock.fragment import Fragment
+
+from xblockutils.resources import ResourceLoader
+from xblockutils.studio_editable import StudioEditableXBlockMixin
+
+from .exceptions import LtiError
+from .oauth import log_authorization_header
+from .lti import Lti1Consumer, Lti2Consumer
+from .outcomes import OutcomeService
+
+
+log = logging.getLogger(__name__)
+
+DOCS_ANCHOR_TAG_OPEN = (
+    "<a "
+    "target='_blank' "
+    "href='"
+    "http://edx.readthedocs.org"
+    "/projects/open-edx-building-and-running-a-course/en/latest/exercises_tools/lti_component.html"
+    "'>"
+)
+RESULT_SERVICE_SUFFIX_PARSER = re.compile(r"^user/(?P<anon_id>\w+)", re.UNICODE)
+ROLE_MAP = {
+    'student': u'Student',
+    'staff': u'Administrator',
+    'instructor': u'Instructor',
+}
+LTI_PARAMETERS = [
+    'lti_message_type',
+    'lti_version',
+    'resource_link_title',
+    'resource_link_description',
+    'user_image',
+    'lis_person_name_given',
+    'lis_person_name_family',
+    'lis_person_name_full',
+    'lis_person_contact_email_primary',
+    'lis_person_sourcedid',
+    'role_scope_mentor',
+    'context_type',
+    'context_title',
+    'context_label',
+    'launch_presentation_locale',
+    'launch_presentation_document_target',
+    'launch_presentation_css_url',
+    'launch_presentation_width',
+    'launch_presentation_height',
+    'launch_presentation_return_url',
+    'tool_consumer_info_product_family_code',
+    'tool_consumer_info_version',
+    'tool_consumer_instance_guid',
+    'tool_consumer_instance_name',
+    'tool_consumer_instance_description',
+    'tool_consumer_instance_url',
+    'tool_consumer_instance_contact_email',
+]
+
+
+LaunchTargetOption = namedtuple('LaunchTargetOption', ['display_name', 'value'])
```

Comment:
For better readability, I suggest moving this line down to just before the definition of the `LaunchTarget` class.